File: sdks/python/src/autonoma/fingerprint.py
"""Deterministic SHA256-based fingerprinting of scenario definitions."""

from __future__ import annotations

import hashlib
import json
from typing import Any


def fingerprint(value: Any) -> str:
    """Compute a 16-char hex fingerprint of any JSON-serializable value."""
    normalized: Any = _sort_keys(value)
    json_str: str = json.dumps(normalized, separators=(",", ":"))
    return hashlib.sha256(json_str.encode()).hexdigest()[:16]


def _sort_keys(obj: Any) -> Any:
    """Recursively sort object keys for deterministic serialization."""
    if isinstance(obj, dict):
        return {k: _sort_keys(v) for k, v in sorted(obj.items())}
    if isinstance(obj, list):
        return [_sort_keys(v) for v in obj]
    return obj

File: sdks/ruby/lib/autonoma/fingerprint.rb
# frozen_string_literal: true

require "openssl"
require "json"

module Autonoma
  module Fingerprint
    # Compute a 16-char hex fingerprint of any JSON-serializable value.
    def self.fingerprint(value)
      normalized = sort_keys(value)
      json_str = JSON.generate(normalized)
      OpenSSL::Digest::SHA256.hexdigest(json_str)[0, 16]
    end

    def self.sort_keys(obj)
      case obj
      when Hash
        obj.sort.to_h.transform_values { |v| sort_keys(v) }
      when Array
        obj.map { |v| sort_keys(v) }
      else
        obj
      end
    end

    private_class_method :sort_keys
  end
end
